### PR TITLE
Add EC2 IMDS v2 support

### PR DESCRIPTION
**Pull request by prognant**
### What does this PR do?
Add instance metadata service v2 (IMDS v2) support.
AWS official documentation : https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html
The agent will still rely on IMDS v1 first. It will do IMDS v2 request only if unauthenticated (IMDS v1) requests fail, thus the actual agent behaviour will only change if IMDS v1 is explicitly disabled, ensuring compatibility with existing deployments.

### Motivation
Properly support EC2 instances with IMDSv1 disabled.

### Additional Notes
Tests captured on an EC2 live instance with an agent development build : 
Request without token on an IMDSv2 only instance : 
```
GET /latest/meta-data/instance-id HTTP/1.1
Host: 169.254.169.254
User-Agent: Go-http-client/1.1
Accept-Encoding: gzip

HTTP/1.0 401 Unauthorized
Content-Length: 343
Content-Type: text/html
Date: Fri, 07 Feb 2020 13:31:39 GMT
Connection: close
Server: EC2ws

<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
	"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
 <head>
  <title>401 - Unauthorized</title>
 </head>
 <body>
  <h1>401 - Unauthorized</h1>
 </body>
</html>
```
Token request (PUT):
```
PUT /latest/api/token HTTP/1.1
Host: 169.254.169.254
User-Agent: Go-http-client/1.1
Content-Length: 0
X-Aws-Ec2-Metadata-Token-Ttl-Seconds: 60
Accept-Encoding: gzip

HTTP/1.0 200 OK
Content-Length: 56
Content-Type: text/plain
Date: Fri, 07 Feb 2020 13:31:39 GMT
X-Aws-Ec2-Metadata-Token-Ttl-Seconds: 60
Connection: close
Server: EC2ws

AQAAAFKw7LwD9Nn5LmMKwWMp9-xj2CI5EuASEmX6F3_g9F3iWvlFTQ==
````
Request to metadata with token:
```
GET /latest/meta-data/instance-id HTTP/1.1
Host: 169.254.169.254
User-Agent: Go-http-client/1.1
X-Aws-Ec2-Metadata-Token: AQAAAFKw7LwD9Nn5LmMKwWMp9-xj2CI5EuASEmX6F3_g9F3iWvlFTQ==
Accept-Encoding: gzip

HTTP/1.0 200 OK
Accept-Ranges: bytes
Content-Length: 19
Content-Type: text/plain
Date: Fri, 07 Feb 2020 13:31:39 GMT
Last-Modified: Fri, 07 Feb 2020 13:30:01 GMT
X-Aws-Ec2-Metadata-Token-Ttl-Seconds: 60
Connection: close
Server: EC2ws

i-0c583456e1956d03e
```

### CODE REVIEW

File: pkg/util/ec2/ec2.go [hunk 1]
```diff
@@ -158,14 +162,53 @@ func getResponse(url string) (*http.Response, error) {
...
 	return res, nil
 }
 
+func getToken() (string, error) {
+	client := http.Client{
+		Timeout: timeout,
+	}
+
+	req, err := http.NewRequest(http.MethodPut, tokenURL, nil)
+	req.Header.Add("X-aws-ec2-metadata-token-ttl-seconds", "60")
```
Comment (ogaca-dd) [inline pkg/util/ec2/ec2.go]: You may want to check err before adding a header.

File: pkg/util/ec2/ec2.go [hunk 2]
```diff
@@ -144,12 +145,15 @@ func extractClusterName(tags []string) (string, error) {
 	return clusterName, nil
 }
 
-func getResponse(url string) (*http.Response, error) {
+func getResponse(url string, method string, headers map[string]string) (*http.Response, error) {
```
Comment (ogaca-dd) [inline pkg/util/ec2/ec2.go]: What do you think about updating the name of the function as it can perform put request now?

**Review by ogaca-dd — COMMENTED** [review]
LGTM just minor comments.

File: pkg/util/ec2/ec2.go [hunk 3]
```diff
@@ -158,14 +162,53 @@ func getResponse(url string) (*http.Response, error) {
 	if err != nil {
 		return nil, err
 	}
-
-	if res.StatusCode != 200 {
+	_, tokenAlreadySet := headers["X-aws-ec2-metadata-token"]
+	if res.StatusCode == 401 && !tokenAlreadySet {
```
[the remark this answers is not in the document]
Comment (prognant) [reply pkg/util/ec2/ec2.go]: Good point, I don't know what the expected behaviour from the metadata service in such case. I will test & check AWS doc to handle that case.

File: pkg/util/ec2/ec2.go [hunk 4]
```diff
@@ -158,14 +162,53 @@ func getResponse(url string) (*http.Response, error) {
...
+		// Retry with a token
+		token, err := getToken()
+		if err != nil {
+			fmt.Println("3")
```
[the remark this answers is not in the document]
Comment (prognant) [reply pkg/util/ec2/ec2.go]: Indeed, to be removed.

File: pkg/util/ec2/ec2.go [hunk 5]
```diff
@@ -144,12 +145,15 @@ func extractClusterName(tags []string) (string, error) {
...
 
-func getResponse(url string) (*http.Response, error) {
+func getResponse(url string, method string, headers map[string]string) (*http.Response, error) {
 	client := http.Client{
 		Timeout: timeout,
 	}
 
-	req, err := http.NewRequest("GET", url, nil)
+	req, err := http.NewRequest(method, url, nil)
+	for header, value := range headers {
```
[the remark this answers is not in the document]
Comment (prognant) [reply pkg/util/ec2/ec2.go]: It makes sense, will do.

File: pkg/util/ec2/ec2.go (hunk 1 above)
Comment (prognant) [reply pkg/util/ec2/ec2.go]: Indeed, will do.

File: pkg/util/ec2/ec2.go (hunk 2 above)
Comment (prognant) [reply pkg/util/ec2/ec2.go]: I'm open to suggestion, I though about `doRequest` not sure it is really better that `getResponse` (in some extend we still get a <method> response in all situation 🤔)

File: pkg/util/ec2/ec2.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (prognant) [reply pkg/util/ec2/ec2.go]: Expired token request leads to a 401 :

```
$ TOKEN=`curl -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 2"` && sleep 3  && curl -v -H "X-aws-ec2-metadata-token: $TOKEN" -v http://169.254.169.254/latest/meta-data/
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    56  100    56    0     0  28000      0 --:--:-- --:--:-- --:--:-- 28000
*   Trying 169.254.169.254:80...
* TCP_NODELAY set
* Connected to 169.254.169.254 (169.254.169.254) port 80 (#0)
> GET /latest/meta-data/ HTTP/1.1
> Host: 169.254.169.254
> User-Agent: curl/7.66.0
> Accept: */*
> X-aws-ec2-metadata-token: AQAAAAgLKK5KPf12uzlfPPj8brc8hufHdPfKYGbNYKUd5ljGtPlPjA==
>
* Mark bundle as not supporting multiuse
* HTTP 1.0, assume close after body
< HTTP/1.0 401 Unauthorized
< Content-Length: 343
< Content-Type: text/html
< Date: Fri, 14 Feb 2020 13:34:17 GMT
< Connection: close
< Server: EC2ws
<
<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
	"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
 <head>
  <title>401 - Unauthorized</title>
 </head>
 <body>
  <h1>401 - Unauthorized</h1>
 </body>
</html>
* Closing connection 0
```

File: pkg/util/ec2/ec2.go (hunk 2 above)
Comment (prognant) [reply pkg/util/ec2/ec2.go]: I will go with doHTTPRequest(...)

**Review by ogaca-dd — APPROVED** [review]
:+1:

**Comment by ishug86** [comment]
@prognant please correct me if I am wrong, I was just curious about this change. Looks like you first try to fetch the info from v1 and then defaults to v2. 
Should it not be reversed? In what case the code will encounter `401` and go to v2?

**Comment by prognant** [comment]
It could be reversed, but the idea was to keep the same behaviour than before the change (i.e. doing IMDSv1)  where possible and support deployment with IMDSv2 only/IMDSv1 disabled. Not changing the default behaviour let some margin to ensure the IMDSv2 implementation is trouble-free.

If we think forward we could plan to do IMDSv2 only because :
* IMDSv2 is enabled everywhere (but optional, i.e. IMDSv1 still works) by default
* IMDSv2 can be mandatory by AWS configuration
* IMDSv1 alone is not possible
So strictly speaking IMDSv1 support is not required anymore and is likely to be dropped at some point .

**Comment by ishug86** [comment]
Thanks @prognant for the explanation, and sorry for responding so late.
As far as I know, IMDS V2 will always be enabled on all instances as its on AWS backend API (on the hypervisor). 
However, With this implementation, the problem could be when someone disable v1 on their machine, it will lead to some call failures for v1  and also not recommended from security perspective as it may confuse folks if they want to disable IMDS V1 on the instance running datadog agent.

**Comment by prognant** [comment]
Thanks for your feedback @ishug86. We understand that IMDSv2 will always be there, we agree that it makes sense to favor IMDSv2 regarding security. Thus we will work on allowing the agent to query IMDSv2 by default in a future release

**Comment by ishug86** [comment]
Thanks @prognant , I will look forward and track the new release.

**Pull request merged**